SerialPrograms/Source/PokemonSV/Programs/TeraRaids/PokemonSV_TeraRoller.cpp

```diff
@@ -0,0 +1,288 @@
+        if (stars == 0){
+            dump_image(env.logger(), env.program_info(), "ReadStarsFailed", *screen.frame);
+        }
+
+        VideoOverlaySet overlay_set(env.console);
+
+        TeraSilhouetteReader silhouette_reader;
+        silhouette_reader.make_overlays(overlay_set);
+        ImageMatch::ImageMatchResult silhouette = silhouette_reader.read(screen);
+        silhouette.log(env.logger(), 100);
+        std::string best_silhouette = silhouette.results.empty() ? "UnknownSilhouette" : silhouette.results.begin()->second;
+        if (silhouette.results.empty()){
+            dump_image(env.logger(), env.program_info(), "ReadSilhouetteFailed", *screen.frame);
+        }
+        else if (PreloadSettings::debug().IMAGE_TEMPLATE_MATCHING){
+            dump_debug_image(env.logger(), "PokemonSV/TeraRoller/" + best_silhouette, "", screen);
+        }
+
+        TeraTypeReader type_reader;
+        type_reader.make_overlays(overlay_set);
+        ImageMatch::ImageMatchResult type = type_reader.read(screen);
+        type.log(env.logger(), 100);
+        std::string best_type = type.results.empty() ? "UnknownType" : type.results.begin()->second;
+        if (type.results.empty()){
+            dump_image(env.logger(), env.program_info(), "ReadTypeFailed", *screen.frame);
+        }
+        else if (PreloadSettings::debug().IMAGE_TEMPLATE_MATCHING){
+            dump_debug_image(env.logger(), "PokemonSV/TeraRoller/" + best_type, "", screen);
+        }
+
+        {
+            std::string log = "Detected a " + std::to_string(stars) + "* " + best_type + " " + best_silhouette;
+            env.console.overlay().add_log(log, COLOR_GREEN);
+            env.log(log);
+        }
+
+        if (!FILTER.should_battle(stars, best_silhouette)) {
+            env.log("Skipping raid...", COLOR_ORANGE);
+            stats.m_skipped++;
+            close_raid(env.program_info(), env.console, context);
+            continue;
+        }
+
+        // Enter tera raid battle alone
+        pbf_press_dpad(context, DPAD_DOWN, 10, 10);
+        pbf_mash_button(context, BUTTON_A, 250);
+        context.wait_for_all_requests();
+        overlay_set.clear();
+        env.console.log("Entering tera raid...");
+        env.console.overlay().add_log("Entering tera raid...", COLOR_WHITE);
+
+        // Run away from the tera raid battle
+        run_from_tera_battle(env.program_info(), env.console, context);
+        context.wait_for_all_requests();
+
+        env.console.log("Checking if tera raid is shiny...");
+        env.console.overlay().add_log("Checking Pokédex...", COLOR_WHITE);
+        open_pokedex_from_overworld(env.program_info(), env.console, context);
+        open_recently_battled_from_pokedex(env.program_info(), env.console, context);
+
+        pbf_wait(context, 200);
+
+        // Since encountering the same species within 5 encounters is possible,
+        // loop through all 5 candidates of recently battled pokemon for shinies

```

Comment:
So if you see the same species, they get grouped together and not bumped to the top of the list?

Reply:
Yes if you encounter something that's already within the 5 Recently Battled Pokémon, the order of the list will not change but the Pokédex page will reflect the more recent encounter.

An example to outline this behavior:
Current Pokédex order: Orthworm, Pawmot, Donphan-Male, Ditto, Noivern
After a new encounter with Donphan-Female: Orthworm, Pawmot, Donphan-Female, Ditto, Noivern

Maybe this can be an option for users to pick if they only want to check only the first in the list to save time or be safer and check all five?

Reply:
This behavior can be useful on single species hunts once the species filter gets added, since you can control it so the species being hunted will always be the first in the list.